### ISSUE TITLE
Strip the prefix `v` from tags created on release event

### ISSUE THREAD

**Issue by Nosfistis** [docker/metadata-action#422]
Following-up on this, I am finding that on the release published event the `v` from the tag version is not being stripped from the `{{version}}` variable. This seems incorrect, if the action is supposed to work correctly also on that event type.

Is this something that was done by design, or instead something that could get updated to match (partially, I agree with the caveats regarding pre-releases) the behaviour on the tag push event?

_Originally posted by @esune in https://github.com/docker/metadata-action/issues/240#issuecomment-1578947442_
            

I realize that this still occurs, and the generated tags contain the `v` prefix.

**Comment by polarathene** [comment]
**TL;DR:** Everything is working correctly as documented. You have misconfigured.

Worth noting from the issue you referenced (_and the one it references_), there is acknowledgement that there is no `release` event support on this actions docs. It's only "supported" in the sense that the event itself results in the `refs/tags/<your tag>` event metadata, which if you have a compatible tag rule to match on that it'll be used.

---

The `v` stripping only occurs during the [`type=semver,pattern=` tag rule](https://github.com/docker/metadata-action#typesemver):

https://github.com/docker/metadata-action/blob/8e1d5461f02b7886d3c1a774bfbd873650445aa2/src/meta.ts#L172-L184

**NOTE:** It's technically discouraged since SemVer 2.0.0 to prefix with `v` (_but IIRC this is still expected for Go tooling, and this action supports the convention implicitly through the `semver` package_):

> A leading "=" or "v" character is stripped off and ignored. Support for stripping a leading "v" is kept for compatibility with v1.0.0 of the SemVer specification but should not be used anymore. [Source](https://www.npmjs.com/package/semver#Versions)

Using the [`release` event type to trigger does set `GITHUB_REF` to `/refs/tags/<your tag>`](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#release) which is required for this semver tag type:

https://github.com/docker/metadata-action/blob/8e1d5461f02b7886d3c1a774bfbd873650445aa2/src/meta.ts#L155-L158

But if you have other tag rules like in the referenced issue:

```yaml
tags: |
  type=ref,event=branch
  type=ref,event=tag
```

That calls different methods to process what tags to generate:

https://github.com/docker/metadata-action/blob/8e1d5461f02b7886d3c1a774bfbd873650445aa2/src/meta.ts#L78-L87

https://github.com/docker/metadata-action/blob/8e1d5461f02b7886d3c1a774bfbd873650445aa2/src/meta.ts#L272-L286

As you can see the `event=tag` will trigger that which just publishes the tag as-is. No `v` would be stripped then.

---

You should be able to [use either of the `on.release` or `on.push.tags` event triggers](https://github.com/docker/metadata-action/issues/168#issue-1148253056) since all that matters here is the `/ref/tags/<your tag>` part of the event, and matching that to the appropriate tag rule.

As shown earlier, there are only three event types actually supported for `type=ref`:

https://github.com/docker/metadata-action/blob/8e1d5461f02b7886d3c1a774bfbd873650445aa2/src/tag.ts#L15-L19

Thus there is no `type=ref,event=release`, and neither you nor the comment you referenced shared what `tags` you configured for the action. You likely either had none and relied upon defaults (_which are documented as not stripping the `v` prefix_), or you copy/pasted the `tags` config from somewhere without understanding them properly.

**Comment by polarathene** [comment]
> I am not sure why pre-releases would behave differently for prefixes though.

https://github.com/docker/metadata-action/blob/902fa8ec7d6ecbf8d84d538b9b233a880e428804/src/meta.ts#L156-L169

https://github.com/docker/metadata-action/blob/8e1d5461f02b7886d3c1a774bfbd873650445aa2/src/meta.ts#L171-L184

> _`loose`: Be more forgiving about not-quite-valid semver strings. (Any resulting output will always be 100% strict compliant, of course.) For backwards compatibility reasons, if the options argument is a boolean value instead of an object, it is interpreted to be the loose param._
> 
> _Strict-mode Comparators and Ranges will be strict about the SemVer strings that they parse._
> - _`valid(v)`: Return the parsed version, or `null` if it's not valid._
> - _`major(v)`: Return the major version number._
> - _`minor(v)`: Return the minor version number._
> - _`patch(v)`: Return the patch version number._
> - _`parse(v)`: Attempt to parse a string as a semantic version, returning either a SemVer object or `null`._
>
> [Source (`semver` NPM docs)](https://www.npmjs.com/package/semver#Functions)

That first snippet I reference calls `semver.valid()`, which if done so as strict-mode presumably emits that warning for pre-release with `v`? Otherwise if it's not occurring there, we have that 2nd snippet which has a pre-release conditional branch..

Looks like it'd probably get treated the same as a non-prerelease, although I'm not sure where `{{version}}` is handled, as those `semver` docs don't mention a `version()` function. From what I can see handlebars receives the template to generate the tag string from, and it's given `sver` as an object with functions on it to call from the template string.

If the `v` prefix stripping occurs somewhere else, I'm not quite sure where 🤔 I assume it's reliant upon `semver` package, so if that doesn't support stripping away the `v` prefix, the bug fix would either be handled up there or some additional config option for using `semver` in the referenced snippets above.

**Comment by esune** [comment]
I am pretty positive I was using  ```type=semver,pattern={{version}}``` and ```type=semver,pattern={{major}}.{{minor}}``` in my action. There also is ```type=ref,event=pr``` as the tagging happens inside of a build action that is reused in a couple workflows, and the tags might be different based on that (see [here](https://github.com/bcgov/traction/blob/a50e9ec5667474d16d4954b2c697559b0363f950/.github/actions/build_acapy/action.yaml#L58-L67) if curious).

I would have to test again and see if the issue persists and what is it exactly, might do it as an experiment next time we need to publish a release by going through a pre-release first).

**Comment by polarathene** [comment]
> There also is `type=ref,event=pr` as the tagging happens inside of a build action that is reused in a couple workflows, and the tags might be different based on that (see [here](https://github.com/bcgov/traction/blob/a50e9ec5667474d16d4954b2c697559b0363f950/.github/actions/build_acapy/action.yaml#L58-L67) if curious).

I went through the [tag history of that repo](https://github.com/bcgov/traction/tags?after=v0.2.13), were these your pre-release style tags in question?:

> ![Image](https://github.com/user-attachments/assets/3d5bc267-b9a1-4c95-b0dc-41b02433d1cd)

Because they were published without a `v` prefix:

> ![Image](https://github.com/user-attachments/assets/f7436173-60a2-47d3-bd30-b08c6f57bf82)

---

> I would have to test again and see if the issue persists and what is it exactly, might do it as an experiment next time we need to publish a release by going through a pre-release first).

Nothing stood out from the other two tag types when I went over the `docker/metadata-action` source, and without a reproduction of it in the wild to reference I'd have to attempt to reproduce myself instead of trying to drive-by assist 😝 

If you do encounter it again, a more minimal workflow to reproduce may also be worthwhile context to confirm it's not due to a more complex interaction.

---

**Off-topic:** Out of curiousity, are you getting any known value from using `type=sha` / ~~`type=ref,event=pr`~~ (_**EDIT:** This one I understand_)?

I don't see much point in the `sha` tags personally:
- Does anyone actually use those tags for your images to reference past images? Generally you'd use an explicit tag and if that tag is prone to being re-assigned you may instead pin by digest at the time.
- It's messy to browse past images that no longer serve a purpose when they're still in the tagged images list 🤷‍♂ (_I'm assuming you chose to do so on the whim that it might be useful before you ever actually had a need for it?_)

<details>
<summary>Ignore (resolved)</summary>

Same with the PR tagging. Not sure if you're aware but secrets on PR triggered workflow runs are not available (_unless using something like `pull_request_target` event and trusting the PR itself cannot exploit access to secrets that way_). IIRC that may not apply to branches on your own repo that PR to another branch, but that doesn't make the feature too useful consistency wise when it comes to third-party contributors. I mention this due to [stuff like this](https://github.com/bcgov/traction/blob/a50e9ec5667474d16d4954b2c697559b0363f950/.github/workflows/on_pr_opened.yaml#L46) which is referencing secrets.

**EDIT:** [Workflow run](https://github.com/bcgov/traction/actions/runs/13272480995/job/37054959975) for [PR](https://github.com/bcgov/traction/pull/1504) on local repo branch. I see you're actually [deploying the PR images with a purpose](https://github.com/bcgov/traction/blob/a50e9ec5667474d16d4954b2c697559b0363f950/.github/workflows/on_pr_opened.yaml#L113-L119) and a conditional to ensure it's [only on local branches](https://github.com/bcgov/traction/blob/a50e9ec5667474d16d4954b2c697559b0363f950/.github/workflows/on_pr_opened.yaml#L21-L33) with a [cleanup step](https://github.com/bcgov/traction/blob/a50e9ec5667474d16d4954b2c697559b0363f950/.github/workflows/on_pr_closed.yaml#L47-L59) (_I was wondering why I didn't see many pr tagged images_ 😅)

</details>

**Comment by esune** [comment]
@polarathene honestly at this point I don't recall anymore, it's been too long 😅 

For the `sha` tag: we use the short sha to deploy to our dev environment and know which image is there at any given time, as it is the target of our CI/CD pipeline. I agree all of those `sha` tags though make the artifact repository messy: it is something that was inherited and we never managed to fully clean-up, we had an action that was supposed to clean up old/unused images (`snok/container-retention-policy`), but after an update it started deleting more than it needed to and we had to turn it off

**Comment by polarathene** [comment]
This issue can be closed.

The following results in the same tags produced by a release event or manual action trigger with equivalent tag value as an input.

```yaml
name: 'Build and Publish the container image'

on:
  release:
    types:
      - published

  workflow_dispatch:
    inputs:
      version:
        description: 'Version to build (e.g., 1.42.0)'
        required: true
        type: string

jobs:
  container:
    runs-on: ubuntu-latest
    permissions:
      contents: read
      packages: write
    steps:
      - name: Prepare image metadata
        id: metadata
        uses: docker/metadata-action@v5
        with:
          images: |
            ghcr.io/${{ github.repository }}
          tags: |
            type=semver,pattern={{major}}.{{minor}}.{{patch}}
            type=semver,pattern={{major}}.{{minor}}
            type=semver,pattern={{major}}

      - name: Set up Docker Buildx
        uses: docker/setup-buildx-action@v3

      - name: Log in to Container Registry
        uses: docker/login-action@v3
        with:
          registry: ghcr.io
          username: ${{ github.repository_owner }}
          password: ${{ secrets.GITHUB_TOKEN }}

      - name: Build and push container image
        uses: docker/build-push-action@v6
        env:
          SOURCE_DATE_EPOCH: 0
        with:
          provenance: false
          push: true
          platforms: linux/amd64,linux/arm64
          tags: ${{ steps.metadata.outputs.tags }}
          outputs: type=image,oci-mediatypes=true,oci-artifact=true,rewrite-timestamp=true
```

I tested with publishing releases with `v1.2.3` with the `latest` and `pre-release` options enabled (_neither option influenced the outcome_), and also a `v1.2.3-alpha` (_pre-release semver, which the action detects and will ignore the major.minor + major release tags_). In all cases the `v` prefix was stripped.

**NOTE:**
- For the `workflow_dispatch` trigger, you would need modify the workflow to explicitly provide the value via appending:
  - `,value=${{ inputs.version }}`
  - `,value=${{ inputs.version || github.ref_name }}` (_compatible fallback that still works with release event_)
- For those that want to prefix the version with a `v`, you could do so with:
  - `type=raw` (_needs explicit value provided, skips any semver logic and thus implicit `latest` tag triggered via default `flavor.latest` setting_)
  - `type=semver,pattern={{ raw }}` (_no stripping of tag value, just needs to be valid semver_)
  - `type=semver,pattern=v{{ major }}.{{ minor }}` (_works but is discouraged_)
  - `type=semver,pattern={{ major }}.{{ minor }},prefix=v` (_better, more consistent/deterministic if pre-release is involved_)
  
  `type=semver` parses the tag into a semver value, which strips that `v`. To retain the `v` prefix, you can prefix to the `pattern` expression like shown above, but this is problematic as explained below. Better to go with `prefix=v`.

  For pre-release detected semver values, the actions logic [only permits `type=semver,pattern={{ raw }}`](https://github.com/docker/metadata-action/blob/ed95091677497158a9ff38b314264cd965388d5e/src/meta.ts#L373-L383) with no customizations in `pattern`, otherwise it will [override any other `pattern` value into `pattern={{ version }}`](https://github.com/docker/metadata-action/blob/ed95091677497158a9ff38b314264cd965388d5e/src/meta.ts#L187-L192).
  - If you use `v` prefixed tags as input, these will only retain the `v` prefix via the raw pattern (_as communicated via the [READEME `type=semver` table](https://github.com/docker/metadata-action?tab=readme-ov-file#typesemver)_).
  - If your tag input lacks a `v` prefix and you want one on the tag only, just use the `prefix=v` option (_better solution overall_).
  - Be mindful of this pre-release behaviour if permitting such semver tags to be published, otherwise `{{ raw }}` usage would retain the `v` prefix, but the "skipped" major and major+minor patterns would be `{{ version }}` patterns that strip the `v` prefix, resulting in two tags. As stated it's better to use `prefix=v` instead.
